### PR TITLE
[WIP] Factor out nn impls using TupleLists and the GetTensors trait.

### LINKED ISSUE #435: Parameter Count
When training large models, it's quite useful to know the size of the model being trained. In pytorch it's fairly straightforward to calculate this:
```python
params = sum(p.numel() for p in model.parameters() if (p.requires_grad or not trainable_only))
```
This is doable because we can use the `model.parameters()` to get a list of all the `Parameter`s in the model, and then sum over the trainable ones.

Currently in dfdx I made a trait for counting parameters called `CountParams`:
```rust
pub trait CountParams {
    fn param_count(&self) -> usize;
}
```
which gets implemented on all tensors like so:
```rust
impl <
    S: Shape,
    E: Dtype + Unit,
    T,
    D: DeviceStorage + ZerosTensor<E> + CopySlice<E>
> CountParams for Tensor<S, E, D, T> {
    fn param_count(&self) -> usize {
        self.shape().num_elements()
    }
}
```

Of course the straightforward way to use this is to implement it on all modules, and then just call `.param_count()` on your model. I don't really think this is a great way to do it though. It seems for every new piece of functionality, we need to constantly implement new traits for each and every module, and as the number of modules grows, this becomes harder and harder to maintain.

One alternative might be to have a state-dict style structure like pytorch has. Something like:
```rust
trait ToStateDict {
    fn state_dict(&self) -> HashMap<String, &Tensor>;
}
```
which would allow modules to specify their storage layout. Once implemented, this could give us CountParams, as well as SaveToNpz for free, because we now have the data and the paths for each param.If the state_dict returned `HashMap<String, &mut Tensor>` instead, it would also give us LoadFromNpz for free as well, because we have the paths as well as the tensors to load into.

Not sure this is the best way, but I don't think constantly implementing new traits will be sustainable long-term.

### PR DESCRIPTION
**Pull request by nkoppel**
A draft of a solution to the issues raised on #435 using TupleLists, which are nested tuples of the form `(T1, (T2, (T3, ())))`. This data structure is fairly simple to serialize modules into and to perform operations on without sacrificing too much performance or sacrificing type-checking at compile time. TupleLists can also be zipped, much like iterators, allowing operations between multiple Modules of the same type to be implemented.

While this will definitely be good for internal code, I am not super happy how GetTensors exposes users who implement custom modules to TupleLists, as specifying the type of a TupleList as the result of appending other TupleLists is pretty unwieldy right now, with `<<Linear<2, 3, f32, TestDevice> as GetTensors>::Tensors as TupleList>::Appended<<Linear<3, 5, f32, TestDevice> as GetTensors>::Tensors>` being the type definition for simply appending two linear layers.

Closes #435, and may make #400 and #425 easier to implement.

Tasks:

- [ ] Implement `GetTensors` for all Modules (`Repeated` may need special support)
- [ ] Implement `GradientUpdate` using `GetTensors`
- [ ] Implement `ResetParams` using `GetTensors` (`linear` and `conv`/`conv2d` will present some issues)
- [ ] Implement `SaveToNpz` using `GetTensors`
- [ ] Implement `LoadFromNpz` using `GetTensors`

### CODE REVIEW

**Comment by coreylowman** [comment]
Yeah this feels quite complex, and given that there may be special cases needed for certain modules I'm not convinced it will necessarily be good for internal code.

I could see EMA being implemented fairly straightforwardly with this, but this wouldn't help gradient accum unless we move gradients to use this tuple list set up.

**Comment by nkoppel** [comment]
Yeah, I have to agree that this approach is over-complicated in practice. TupleLists also don't really feel like they belong in dfdx, and implementing the GetTensors trait for end users is unwieldy. It has also been very difficult to implement GradientUpdate using TupleFold due to issues with lifetimes.
